deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
         pass
 
 
-def _add_default_tags(tags=None):
-    """ add default tags to stats """
-    if not __version__:
-        # defensive case, should not be hit
-        return tags
-    if __version__ and not tags:
-        tags = dict()
-    tags['deploy_agent_version'] = __version__
-    return tags
+def create_stats_timer(name, sample_rate=1.0, tags=None):
+    if IS_PINTEREST:
+        from pinstatsd.statsd import statsd_context_timer
 
+        class TimingOnlyStatClient:
+            """ timing only stat client in order to use caching """
+            @staticmethod
+            def timing(*args, **kwargs):
+                client = MetricClient()
+                return client.send_context_timer(*args, **kwargs)
 
-def create_stats_timer(stats, sample_rate=1.0, tags=None):
-    if IS_PINTEREST:
-        try:
-            from pinstatsd.statsd import statsd_context_timer
-            tags = _add_default_tags(tags)
-            return statsd_context_timer(entry_name=stats, sample_rate=sample_rate, tags=tags)
-        except Exception as e:
-            error_msg = str(e)
-            log.error('unable to send metric: {}'.format(error_msg))
+        timer = statsd_context_timer(entry_name=name, sample_rate=sample_rate, tags=tags)
+        timer._statsd_context_timer__stat_client = TimingOnlyStatClient()
+        return timer
     else:
-        return DefaultStatsdTimer()
+        return
 
 
-def create_sc_timing(stat, value, sample_rate=1.0, tags=None):
+def create_sc_timing(name, value, sample_rate=1.0, tags=None):
     if IS_PINTEREST:
-        try:
-            from pinstatsd.statsd import sc
-            tags = _add_default_tags(tags)
-            return sc.timing(stat, value, sample_rate=sample_rate, tags=tags)
-        except Exception as e:
-            error_msg = str(e)
-            log.error('unable to send metric: {}'.format(error_msg))
+        mtype = 'timing'
```

Comment:
all 3 create_  methods have basically same code. Can we parameterize mtype and cleanup?

Reply:
I tried this change but felt the resulting code was not as readable as keeping the original structure.  It also changed all the calls.  Let me know if you really want it!

Reply:
I don't understand why this methods should even exist after introducing MetricsClient.

Reply:
These are functions that are called by the existing code.  I kept the entry points in place to avoid refactoring too much of the downstream code.